irohad/ametsuchi/impl/postgres_query_executor.cpp

```diff
@@ -151,64 +162,76 @@ namespace iroha {
+          | boost::adaptors::filtered(pred);
+      boost::transform(filtered_txs,
+                       std::back_inserter(result),
+                       [&](const auto &tx) { return clone(tx); });
 
       return result;
     }
 
-    template <typename Q, typename P, typename F, typename B>
-    QueryResponseBuilderDone PostgresQueryExecutorVisitor::executeQuery(F &&f,
-                                                                        B &&b) {
-      using T = concat<Q, P>;
+    template <typename QueryTuple,
+              typename PermissionTuple,
+              typename QueryExecutor,
+              typename ResponseCreator,
+              typename ErrResponse>
+    QueryExecutorResult PostgresQueryExecutorVisitor::executeQuery(
+        QueryExecutor &&query_executor,
+        ResponseCreator &&response_creator,
+        ErrResponse &&err_response) {
+      using T = concat<QueryTuple, PermissionTuple>;
       try {
-        soci::rowset<T> st = std::forward<F>(f)();
+        soci::rowset<T> st = std::forward<QueryExecutor>(query_executor)();
         auto range = boost::make_iterator_range(st.begin(), st.end());
 
         return apply(
-            viewPermissions<P>(range.front()), [range, &b](auto... perms) {
+            viewPermissions<PermissionTuple>(range.front()),
+            [this,
+             range,
+             &response_creator,
+             err_response =
+                 std::forward<ErrResponse>(err_response)](auto... perms) {
               bool temp[] = {not perms...};
               if (std::all_of(std::begin(temp), std::end(temp), [](auto b) {
                     return b;
                   })) {
-                return stateful_failed;
+                return query_response_factory_->createErrorQueryResponse(
+                    QueryErrorType::kStatefulFailed,
+                    err_response(),
+                    this->query_hash_);
               }
               auto query_range = range
                   | boost::adaptors::transformed([](auto &t) {
-                                   return rebind(viewQuery<Q>(t));
+                                   return rebind(viewQuery<QueryTuple>(t));
                                  })
                   | boost::adaptors::filtered([](const auto &t) {
                                    return static_cast<bool>(t);
                                  })
                   | boost::adaptors::transformed([](auto t) { return *t; });
-              return std::forward<B>(b)(query_range, perms...);
+              return std::forward<ResponseCreator>(response_creator)(
+                  query_range, perms...);
             });
       } catch (const std::exception &e) {
         log_->error("Failed to execute query: {}", e.what());
-        return stateful_failed;
+        return query_response_factory_->createErrorQueryResponse(
+            QueryErrorType::kStatefulFailed,
+            "failed to execute query: " + std::string(e.what()),
```

Comment:
Replication of strings. Also you seem to log some errors, but not others. Is this intentional?
If you want to log all errors you can wrap logging and error construction in a function to reduce code duplication.